engine/access/rpc/backend/backend_network.go

```diff
@@ -23,108 +20,100 @@ type backendNetwork struct {
+//
+// The block may or may not be finalized in the future; the client can retry later.
+func (b *backendNetwork) GetProtocolStateSnapshotByBlockID(ctx context.Context, blockID flow.Identifier) ([]byte, error) {
 	snapshot := b.state.AtBlockID(blockID)
 	snapshotHeadByBlockId, err := snapshot.Head()
 	if err != nil {
-		if errors.Is(err, state.ErrUnknownSnapshotReference) {
-			return nil, status.Errorf(codes.NotFound, "failed to get a valid snapshot: block not found")
-		}
-		return nil, status.Errorf(codes.Internal, "could not get header by blockID: %v", err)
+		// storage.ErrNotFound is specifically NOT allowed since the snapshot's reference block must exist
+		// within the snapshot.
+		err = access.RequireErrorIs(ctx, err, state.ErrUnknownSnapshotReference)
+		return nil, access.NewDataNotFoundError("snapshot", fmt.Errorf("failed to retrieve a valid snapshot: block not found"))
 	}
 
 	// Because there is no index from block ID to finalized height, we separately look up the finalized
 	// block ID by the height of the queried block, then compare the queried ID to the finalized ID.
 	// If they match, then the queried block must be finalized.
 	blockIDFinalizedAtHeight, err := b.headers.BlockIDByHeight(snapshotHeadByBlockId.Height)
 	if err != nil {
-		if errors.Is(err, storage.ErrNotFound) {
-			// The block exists, but no block has been finalized at its height. Therefore, this block
-			// may be finalized in the future, and the client can retry.
-			return nil, status.Errorf(codes.FailedPrecondition,
-				"failed to retrieve snapshot for block with height %d: block not finalized and is above finalized height",
-				snapshotHeadByBlockId.Height)
-		}
-		return nil, status.Errorf(codes.Internal, "failed to lookup block id by height %d", snapshotHeadByBlockId.Height)
+		err = access.RequireErrorIs(ctx, err, storage.ErrNotFound)
+
+		// The block exists, but no block has been finalized at its height. Therefore, this block
+		// may be finalized in the future, and the client can retry.
+		return nil, access.NewPreconditionFailedError(
+			fmt.Errorf("failed to retrieve snapshot for block with height %d: block not finalized and is above finalized height",
+				snapshotHeadByBlockId.Height))
 	}
 
 	if blockIDFinalizedAtHeight != blockID {
 		// A different block than what was queried has been finalized at this height.
 		// Therefore, the queried block will never be finalized.
-		return nil, status.Errorf(codes.InvalidArgument,
-			"failed to retrieve snapshot for block: block not finalized and is below finalized height")
+		return nil, access.NewInvalidRequestError(fmt.Errorf("failed to retrieve snapshot for block: block not finalized and is below finalized height"))
 	}
 
 	data, err := convert.SnapshotToBytes(snapshot)
 	if err != nil {
-		return nil, status.Errorf(codes.Internal, "failed to convert snapshot to bytes: %v", err)
+		return nil, access.RequireNoError(ctx, fmt.Errorf("failed to convert snapshot to bytes: %w", err))
```

Comment:
Only if we can guarantee that we are _always_ be able to generate a snapshot for any block (including the root block when the node is freshly bootstrapped) we can use ` access.RequireNoError` here (same as [comment above](url))

```suggestion
		err = access.RequireErrorIs(ctx, err, protocol.ErrSealingSegmentBelowRootBlock, protocol.NewUnfinalizedSealingSegmentErrorf(""))
		return nil, fmt.Errorf("snapshots might not be possible for every block: %w", err)
```